src/Parallel/PhaseControl/PhaseControlTags.hpp

```diff
@@ -23,77 +23,65 @@ namespace PhaseControl {
           using tag = typename decltype(tag_v)::type;
-          tuples::get<tag>(current_state) = typename tmpl::at<
-              TagsAndCombines,
-              tmpl::index_if<
-                  TagsAndCombines,
-                  std::is_same<tmpl::pin<tag>, detail::get_tag<tmpl::_1>>>>::
-              combine_method{}(tuples::get<tag>(current_state),
-                               tuples::get<tag>(element));
+          tuples::get<tag>(current_state) = typename tag::combine_method{}(
+              tuples::get<tag>(current_state), tuples::get<tag>(element));
         });
     return current_state;
   }
 };
 
-template <typename TagsAndCombines>
+/// A flexible combine invokable that combines into a `tuples::TaggedTuple` a
+/// new `tuples::TaggedTuple` with a subset of the original tags, and combines
+/// according to type aliases `main_combination_method`s that are required to be
+/// defined in each tag.
+///
+/// \note This is _not_ usable with charm++ reductions; it mutates the current
+/// state in-place. This is constructed for the use-case where the main chare
+/// stores a persistent data structure and combines reduction data as it arrives
+/// from the other chares.
 struct TaggedTupleMainCombine {
   template <typename... CurrentTags, typename... CombineTags>
-  tuples::TaggedTuple<CurrentTags...> operator()(
-      tuples::TaggedTuple<CurrentTags...> current_state,
+  static void apply(
+      const gsl::not_null<tuples::TaggedTuple<CurrentTags...>*> current_state,
       const tuples::TaggedTuple<CombineTags...>& element) {
     tmpl::for_each<tmpl::list<CombineTags...>>(
         [&current_state, &element](auto tag_v) noexcept {
           using tag = typename decltype(tag_v)::type;
-          tuples::get<tag>(current_state) = typename tmpl::at<
-              TagsAndCombines,
-              tmpl::index_if<
-                  TagsAndCombines,
-                  std::is_same<tmpl::pin<tag>, detail::get_tag<tmpl::_1>>>>::
-              main_combine_method{}(tuples::get<tag>(current_state),
-                                    tuples::get<tag>(element));
+          tuples::get<tag>(*current_state) =
+              typename tag::main_combine_method{}(
+                  tuples::get<tag>(*current_state), tuples::get<tag>(element));
         });
-    return current_state;
   }
 };
-// @}
 
 /// A `Parallel::ReductionData` with a single `Parallel::ReductionDatum` for a
 /// given tagged tuple type determined by `TagsPresent`, and performs the
 /// combine according to `TagsAndCombines`, which must be a `tmpl::list` of
 /// `PhaseControl::TagAndCombine`s.
-template <typename TagsPresent, typename TagsAndCombines>
-using reduction_data = Parallel::ReductionData<
-    Parallel::ReductionDatum<tuples::tagged_tuple_from_typelist<TagsPresent>,
-                             TaggedTupleCombine<TagsAndCombines>>>;
+///
+/// Each tag in the `TagsPresent` may either be a `TagsAndCombines` or otherise
```

Comment:
there appears to be some renaming of either the dox or the type aliases that caused them to be out of sync